cassiopeia/data/CassiopeiaTree.py

```diff
@@ -186,26 +209,36 @@ def populate_tree(self, tree: Union[str, ete3.Tree, nx.DiGraph]) -> None:
             )
 
     def __check_network_initialized(self) -> None:
+        """Checks that topology has been initialized."""
         if self.__network is None:
             raise CassiopeiaTreeError("Tree has not been initialized.")
 
-    def set_character_matrix(self, character_matrix: pd.DataFrame):
-        """Initializes a character matrix in the object."""
+    def set_character_matrix(
+        self, character_matrix: pd.DataFrame, set_character_states: bool = False
+    ):
+        """Initializes a character matrix in the object.
+        
+        Args:
+            character_matrix: Character matrix of mutation observations.
+            set_character_states: Set character states at the leaves of
+                the tree topology, if it exists.
```

Comment:
This point, related to the one above, is fantastic. My intention is that a user can use _any_ character matrix of the tree for downstream analysis (which I think is clear from the motivation of creating the `Layers` class). And to a degree you're right, `character_matrix` does represent `current_character_matrix`, but I don't necessarily want to overwrite the existing `character_matrix` attribute every time you reset the leaves of the tree with new character states. In the same way, the desired behavior here was to not necessarily rewrite the states of the leaves when you set a new default character matrix.

I think this would be more clear though if we decouple setting the leaves of the tree from the character matrix setting. This would mean that a user would have to call `initialize_character_states_at_leaves` after setting the `character_matrix` attribute - something like:

```
tree.character_matrix = <character_matrix>
tree.initialize_character_states_at_leaves()
``` 

Or, if there is another layer available:

```
tree.set_character_states_at_leaves(layer='modified')
```

What do you think of this behavior?

Reply:
I actually really like your suggestion. I always think being explicit (and doing less things implicitly, like setting the character states at the leaves when setting the character matrix) is better. Why don't we go ahead and decouple the character matrix with the character states at the leaves?

Reply:
I agree - this all sounds great! I'm trying to decide if it would make the most sense to set the attribute `character_matrix` when you call `set_character_states_at_leaves(character_matrix=cm)`. This is why I removed this option initially, I can't decide what the best behavior would be.